xlators/cluster/afr/src/afr-open.c

```diff
@@ -271,34 +294,209 @@ afr_fd_ctx_need_open(fd_t *fd, xlator_t *this, unsigned char *need_open)
+    local->call_count = need_open_count;
+
+    for (i = 0; i < priv->child_count; i++) {
+        if (!local->need_open[i])
+            continue;
+
+        if (IA_IFDIR == local->fd->inode->ia_type) {
+            gf_msg_debug(this->name, 0, "opening fd for dir %s on subvolume %s",
+                         local->loc.path, priv->children[i]->name);
+            STACK_WIND_COOKIE(frame, afr_openfd_fix_open_cbk, (void *)(long)i,
+                              priv->children[i],
+                              priv->children[i]->fops->opendir, &local->loc,
+                              local->fd, NULL);
+        } else {
+            gf_msg_debug(this->name, 0,
+                         "opening fd for file %s on subvolume %s",
+                         local->loc.path, priv->children[i]->name);
+
+            STACK_WIND_COOKIE(
+                frame, afr_openfd_fix_open_cbk, (void *)(long)i,
+                priv->children[i], priv->children[i]->fops->open, &local->loc,
+                local->fd_ctx->flags & ~(O_CREAT | O_EXCL | O_TRUNC), local->fd,
+                NULL);
+        }
+        if (!--need_open_count)
+            break;
+    }
+    return 0;
+
+out:
+    afr_fd_ctx_reset_need_open(local->fd, this, local->need_open);
+    AFR_STACK_DESTROY(frame);
+    return 0;
+}
+
+static int
+afr_is_reopen_allowed_cbk(call_frame_t *frame, void *cookie, xlator_t *this,
+                          int32_t op_ret, int32_t op_errno,
+                          struct gf_flock *lock, dict_t *xdata)
+{
+    afr_local_t *local = frame->local;
+    afr_private_t *priv = NULL;
+    int ret = -1;
+    int call_count = 0;
+    int i = (long)cookie;
+    int32_t fd_reopen_status = -1;
+
+    priv = this->private;
+    local->replies[i].valid = 1;
+    local->replies[i].op_ret = op_ret;
+    local->replies[i].op_errno = op_errno;
+    if (op_ret != 0) {
+        gf_msg(this->name, GF_LOG_ERROR, op_errno, AFR_MSG_LK_HEAL_DOM,
+               "Failed getlk for %s", uuid_utoa(local->fd->inode->gfid));
+    }
+
+    /* Currently we get 3 values from the lower layer (protocol/client) in the
+     * getlk_cbk.
+     *  0 : No conflicting locks are held and reopen is allowed
+     *  1 : Conflicting locks are held and reopen is not allowed
+     *  2 : FD is not valid
+     *
+     * - If we get 1 from any of the bricks, will block the reopen taking this
+     *   as high priority.
+     * - If we get 2 from all the replies, we will not reopen since we do not
+     *   know the correct status.
+     * - If we get 2 from few brick and 1 from one or more bricks, then we
+     *   will block reopen.
+     * - If we get 2 from few bricks and 0 from one or more bricks, then we will
+     *   allow the reopen.
+     *
+     *   We will update the local->fd_reopen_status only when the value returned
+     *   from lower layer is >= 0 and < 2. We will not set 2 in local, since it
+     *   can lead to unexpected behaviours.
+     */
```

Comment:
i.e we might still need to have the current checks of go/no go, but it can be achieved with just with op_ret value of the cbk and whether the key is set in the dict or not.

Reply:
With just 2 value we can not handle all the cases.
Lets consider the change that you have suggested of not calling set_fd_reopen_status() in client3_3_lk & client4_0_lk and 
 reopening the fd in afr only if the op_ret of all valid replies are 0.

Case 1:
When strict-locks setting is disabled and we get EBADFD from all the bricks i.e., op_ret != 0. Here we will block the reopen with the change in afr that it expects fd_reopen_status to be 0 from all the valid replies.

Case 2:
If we consider default value as REOPEN that can handle the previous case, and reopen the fd. But this can lead to reopen in a similar case where strict-locks is enabled, i.e., op_ret != 0 from all the replies. But since we consider default as REOPEN, we will go ahead and reopen the fd, but there can be another fd holding the lock.

There could be more similar cases as well.
What we can do is, as you suggested, instead of checking the dict value inside the LOCK(FRAME->LOCK) and having one more variable local->fd_reopen_status, we can save the xdata in local->replies[child_index].xdata and compare that when call_count becomes 0. And we need to compare this value even when op_ret != 0, so that we know whether strict-locks is enabled or disabled and this status needs to be set in client(3/4)_(3/0)_lk.

Let me know if this makes sense.

Reply:
Ah this makes sense. Yes we need to evaluate replies from all children.  But if are we are defining enum values, then it is better to use them in AFR too instead of 0,1,2 etc.

Reply:
Done.